pkg/kstatus/observe/observer/observer.go

```diff
@@ -0,0 +1,197 @@
+type ObserversFactoryFunc func(reader ObserverReader, mapper meta.RESTMapper) (resourceObservers map[schema.GroupKind]ResourceObserver, defaultObserver ResourceObserver)
+
+// Observer provides functionality for polling a cluster for status for a set of resources.
+type Observer struct {
+	Reader client.Reader
+	Mapper meta.RESTMapper
+
+	AggregatorFactoryFunc AggregatorFactoryFunc
+	ReaderFactoryFunc     ReaderFactoryFunc
+	ObserversFactoryFunc  ObserversFactoryFunc
+}
+
+// Observe will create a new StatusObserverRunner that will poll all the resources provided and report their status
+// back on the event channel returned. The StatusObserverRunner can be cancelled at any time by cancelling the
+// context passed in.
+// If stopOnCompleted is set to true, then the runner will stop observing the resources when the StatusAggregator
+// determines that all resources has been fully reconciled. If this is set to false, the observer will keep running
+// until cancelled.
+// The pollInterval specifies how often the observer should poll the cluster for the latest state of the resources.
+func (s *Observer) Observe(ctx context.Context, identifiers []wait.ResourceIdentifier, pollInterval time.Duration, stopOnCompleted bool) <-chan event.Event {
+	eventChannel := make(chan event.Event)
+
+	go func() {
+		defer close(eventChannel)
+
+		observerReader, err := s.ReaderFactoryFunc(s.Reader, s.Mapper, identifiers)
+		if err != nil {
+			eventChannel <- event.Event{
+				EventType: event.ErrorEvent,
+				Error:     err,
+			}
+			return
+		}
+		observers, defaultObserver := s.ObserversFactoryFunc(observerReader, s.Mapper)
+		aggregator := s.AggregatorFactoryFunc(identifiers)
+
+		runner := &StatusObserverRunner{
+			ctx:                       ctx,
+			reader:                    observerReader,
+			observers:                 observers,
+			defaultObserver:           defaultObserver,
+			identifiers:               identifiers,
+			previousObservedResources: make(map[wait.ResourceIdentifier]*event.ObservedResource),
+			eventChannel:              eventChannel,
+			statusAggregator:          aggregator,
+			stopOnCompleted:           stopOnCompleted,
+			pollingInterval:           pollInterval,
+		}
+		runner.Run()
+	}()
+
+	return eventChannel
+}
+
+// StatusObserverRunner is responsible for polling of a set of resources. Each call to Observe will create
+// a new StatusObserverRunner, which means we can keep state in the runner and all data will only be accessed
+// by a single goroutine, meaning we don't need synchronization.
+// The StatusObserverRunner uses an implementation of the ObserverReader interface to talk to the
+// kubernetes cluster. Currently this can be either the cached ObserverReader that syncs all needed resources
+// with LIST calls before each polling loop, or the normal ObserverReader that just forwards each call
+// to the client.Reader from controller-runtime.
+type StatusObserverRunner struct {
+	ctx context.Context
```

Comment:
plz comment each variable for what it is used for

Reply:
I have added comments here.